go-controller/pkg/node/node.go

```diff
@@ -440,13 +454,192 @@ func (n *OvnNode) Start(wg *sync.WaitGroup) error {
 		}
 	}
 
+	n.WatchCRD()
 	n.WatchEndpoints()
 
 	err = cniServer.Start(cni.HandleCNIRequest)
 
 	return err
 }
 
+func (n *OvnNode) WatchCRD() {
+	n.watchFactory.AddCRDHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			crd := obj.(*apiextension.CustomResourceDefinition)
+			klog.Infof("Adding CRD %s to cluster", crd.Name)
+			// currrently the only reason we watch crds is to determine if the watchers for egressfirewall
+			// should be running
+			if n.egressfirewallEnabled {
+				return
+			}
+			existingCRDs, err := n.watchFactory.GetCRDS()
+			if err != nil {
+				klog.Errorf("Error getting CRDs in the cluster: %v", err)
+				return
+			}
+			isEgressFirewallCRD := false
+			isNodeDNSCRD := false
+			for _, existingCRD := range existingCRDs {
+				if existingCRD.Name == egressfirewallCRD ||
+					crd.Name == egressfirewallCRD {
+					isEgressFirewallCRD = true
+				}
+				if existingCRD.Name == nodeDNSInfoCRD ||
+					crd.Name == nodeDNSInfoCRD {
+					isNodeDNSCRD = true
+				}
+				if isNodeDNSCRD && isEgressFirewallCRD {
+					err := n.watchFactory.InitializeEgressFirewallWatchFactory()
+					if err != nil {
+						klog.Errorf("Error Creating EgressFirewallWatchFactory: %v", err)
+						return
+					}
+					klog.Infof("Enabling EgressFirewall")
+					n.egressfirewallEnabled = true
+					n.egressFirewallHandler = n.WatchEgressFirewall()
+					n.egressFirewallDNS, err = egressfirewalldns.NewEgressDNS(n.name, n.watchFactory, n.Kube, make(chan struct{}))
+					if err != nil {
+						klog.Errorf("Egressfirewall could not start properly on %s", n.name)
+						return
+					}
+					n.egressFirewallDNS.Run(egressFirewallDNSDefaultDuration)
+					return
+				}
+			}
+
+		},
+		UpdateFunc: func(old, newer interface{}) {},
+		DeleteFunc: func(obj interface{}) {
+			crd := obj.(*apiextension.CustomResourceDefinition)
+			klog.Infof("Deleting CRD %s from cluster", crd.Name)
+			// if the egressFirewall is already shutdown we don't care about other crds
+			if !n.egressfirewallEnabled {
+				return
+			}
+
+			existingCRDs, err := n.watchFactory.GetCRDS()
+			if err != nil {
+				klog.Errorf("Error getting CRDs in the cluster: %v", err)
+				return
+			}
+			isEgressFirewallCRD := false
+			isNodeDNSCRD := false
+			for _, existingCRD := range existingCRDs {
+				if existingCRD.Name == egressfirewallCRD ||
+					crd.Name == egressfirewallCRD {
+					isEgressFirewallCRD = true
+				}
+				if existingCRD.Name == nodeDNSInfoCRD ||
+					crd.Name == nodeDNSInfoCRD {
+					isNodeDNSCRD = true
+				}
+				if !isNodeDNSCRD || !isEgressFirewallCRD {
+					klog.Infof("Disabling EgressFirewall")
+					n.egressfirewallEnabled = false
+					n.watchFactory.RemoveEgressFirewallHandler(n.egressFirewallHandler)
+					n.egressFirewallHandler = nil
+					n.egressFirewallDNS.Shutdown()
+					return
+				}
+			}
+		},
+	}, nil)
+
+}
+
+func (n *OvnNode) WatchEgressFirewall() *factory.Handler {
```

Comment:
Rather than going through all the trouble to watch CRDs, why not just make this a config flag?

Reply:
it was determined that the footprint would be so small that it would not matter and not worth cluttering up the cli with another flag https://github.com/ovn-org/ovn-kubernetes/pull/1432#discussion_r457675211